proto/sql_api.proto

```diff
@@ -0,0 +1,154 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Vivek Menezes (vivek@cockroachlabs.com) 
+
+syntax = "proto2";
+package cockroach.proto;
+option go_package = "proto";
+
+import "cockroach/proto/data.proto";
+import "cockroach/proto/errors.proto";
+import "gogoproto/gogo.proto";
+
+option (gogoproto.sizer_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+
+// ClientCmdID provides a unique ID for client commands. Clients which
+// provide ClientCmdID gain operation idempotence. In other words,
+// clients can submit the same command multiple times and always
+// receive the same response. This is common on retries over flaky
+// networks. However, the system imposes a limit on how long
+// idempotence is provided. Retries over an hour old are not
+// guaranteed idempotence and may be executed more than once with
+// potentially different results.
+//
+// ClientCmdID contains the client's timestamp and a client-generated
+// random number. The client Timestamp is specified in unix
+// nanoseconds and is used for some uniqueness but also to provide a
+// rough ordering of requests, useful for data locality on the
+// server. The Random is specified for additional uniqueness.
+// NOTE: An accurate time signal IS NOT required for correctness.
+message ClientCmdID {
+  // Nanoseconds since Unix epoch.
+  optional int64 wall_time = 1 [(gogoproto.nullable) = false];
+  optional int64 random = 2 [(gogoproto.nullable) = false];
+}
+
+// ReadConsistencyType specifies what type of consistency is observed
+// during read operations.
+enum ReadConsistencyType {
+  option (gogoproto.goproto_enum_prefix) = false;
+  // CONSISTENT reads are guaranteed to read committed data; the
+  // mechanism relies on clocks to determine lease expirations.
+  CONSISTENT = 0;
+  // CONSENSUS requires that reads must achieve consensus. This is a
+  // stronger guarantee of consistency than CONSISTENT.
+  //
+  // TODO(spencer): current unimplemented.
+  CONSENSUS = 1;
+  // INCONSISTENT reads return the latest available, committed values.
+  // They are more efficient, but may read stale values as pending
+  // intents are ignored.
+  INCONSISTENT = 2;
+}
+
+// CmdRequestHeader is supplied with every CmdRequest.
+message CmdRequestHeader {
+  // Timestamp specifies time at which read or writes should be
+  // performed. If the timestamp is set to zero value, its value
+  // is initialized to the wall time of the receiving gateway node.
+  optional Timestamp timestamp = 3 [(gogoproto.nullable) = false];
```

Comment:
MySQL has `SET timestamp=NNNN`, although it's not normally used from user code (it's written into the replication logs by the master)